### PR TITLE
array notation to array range transformation (closes #1576)

### LINKED ISSUE #1576: Convert array syntax (without ranges) to ranges and loops. Should there be a canonical form?
The expressions:
`my_array = 2`
`my_array(:,:) = 2`
```
do i = 1, size(my_array, 1)
    do j = 1, size(my_array, 2)
        my_array(i,j) = 2
    end do
end do
```
are all semantically equivalent (*well... with no order implied in the array forms). We have a transformation to transform the 2nd form to the 3rd form, but in some situations we also need to transform the 1st form to the 3rd (maybe by going 1st->2nd->3rd form).

And additional question is that our transfomations search and apply usually to just one of the forms, but we often would want to apply a transformation regardless of how the code it written. We can explore if it would make sense to consider one of them the canonical form and let the fparser_reader already transform it to the proper cannonical format. At least between the 1st and 2nd forms.

### PR DESCRIPTION
**Pull request by rupertford**
Will also be implemented in PSyAD as part of the preprocessing.

### CODE REVIEW

**Comment by rupertford** [comment]
Need to add tests, doc string example for transformation, update documentation and look at dotproduct transformation to see how the bounds are determined - as we currently just add lbound and ubound whereas the dotproduct transformation provides actual values from the array declaration where possible.

**Comment by rupertford** [comment]
Also currently failing for PSyAD with `helmholtz_operator_kernel` (but seems to work for `w2_to_w1_projection_kernel`).

**Comment by rupertford** [comment]
Ready for first review from @arporter or @sergisiso. This transformation returns the bounds of the array range if they can be determined. I think this is the right thing to do, rather than just returning ':'.

**Comment by rupertford** [comment]
The PSyAD part has been dropped as the implementation does not work with the dotproduct transformation, which needs modifying.

File: src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+    '''
+    code = CODE.replace("  real, dimension(10) :: a\n",
+                        "  integer :: n\n  real, dimension(n) :: a\n")
+    result = apply_trans(fortran_reader, fortran_writer, code)
+    assert "a(1:n) = b\n" in result
+
+
+def test_range(fortran_reader, fortran_writer):
+    '''Test that an array slice remains unchanged.'''
```
Comment (arporter) [inline src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py:151]: As mentioned below, I think this really belongs in the validate test.

File: src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    size of the array dimension is unknown.
+
+    '''
+    code = CODE.replace("dimension(10)", "dimension(:)")
+    result = apply_trans(fortran_reader, fortran_writer, code)
+    assert "a(:) = b\n" in result
+
+
+def test_variable(fortran_reader, fortran_writer):
+    '''Test that a array notation gets replaced with an array range when
```
Comment (arporter) [inline src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py:140]: s/a array notation/reference to an array/

File: src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    node = psyir.walk(Reference)[0]
+    symbol = node.symbol
+    lower_bound, upper_bound, step = \
+        ArrayNotation2ArrayRangeTrans._get_array_bound(symbol, 0)
+    assert isinstance(lower_bound, BinaryOperation)
+    assert lower_bound.operator == BinaryOperation.Operator.LBOUND
+    assert isinstance(upper_bound, BinaryOperation)
+    assert upper_bound.operator == BinaryOperation.Operator.UBOUND
+    assert isinstance(step, Literal)
+    assert step.value == "1"
```
Comment (arporter) [inline src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py:108]: For completeness, please add a test for an idx other than zero. As it's an internal method I won't ask for error checking on the value of `idx`.

**Review by arporter — CHANGES_REQUESTED** [review]
Nicely done. I think this transformation needs a better name though and it also needs to be
added to the User Guide.
Apart from that there are just some small tweaks to do.
New test file covers new implementation file.
All tests and examples are OK with compilation.
There is a TODO mentioning #1576 that needs to be looked at:

    ./src/psyclone/psyir/frontend/fparser2.py:        # TODO #1576 this functionality is very similar to that needed in

Reference guide builds with no new warnings.

File: src/psyclone/psyir/transformations/arraynotation2arrayrange_trans.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author R. W. Ford
+
+'''Module providing a transformation from PSyIR Array notation to an
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/transformations/arraynotation2arrayrange_trans.py:36]: Changed the name and updated the description.

File: src/psyclone/psyir/transformations/arraynotation2arrayrange_trans.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author R. W. Ford
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/transformations/arraynotation2arrayrange_trans.py:34]: Oops.

File: src/psyclone/psyir/transformations/arraynotation2arrayrange_trans.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        unknown (a deferred or attribute type) then the LBOUND and
+        UBOUND PSyIR nodes are used.
+
+        :param symbol: the symbol that we are interested in.
+        :type symbol: :py:class:`psyir.symbols.DataSymbol`
+        :param int index: the (array) reference index that we are \
+            interested in.
+
+        :returns: the loop bounds for this array index.
+        :rtype: (Literal, Literal, Literal) or \
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/transformations/arraynotation2arrayrange_trans.py:94]: tuple-ified.

File: src/psyclone/psyir/transformations/arraynotation2arrayrange_trans.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+        '''
+        # pylint: disable=unidiomatic-typecheck
+        if not type(node) is Reference:
+            raise TransformationError(
+                f"The supplied node should be a Reference but found "
+                f"'{type(node).__name__}'.")
+        if not node.symbol.is_array:
+            raise TransformationError(
+                f"The supplied node should be a Reference that references a "
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/transformations/arraynotation2arrayrange_trans.py:138]: Modified.

File: src/psyclone/psyir/transformations/arraynotation2arrayrange_trans.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+                f"symbol that is an array, but '{node.symbol.name}' is not.")
+
+    def apply(self, node, options=None):
+        '''Apply the ArrayNotation2ArrayRangeTrans transformation to the
+        specified node. The node must be a Reference. If the Reference
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/transformations/arraynotation2arrayrange_trans.py:143]: Updated.

File: src/psyclone/psyir/transformations/arraynotation2arrayrange_trans.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+        is to an array then the reference is replaced by an
+        ArrayReference with appropriate explicit range nodes (termed
+        colon notation in Fortran).
+
+        :param node: a Reference node.
+        :type node: :py:class:`psyclone.psyir.nodes.Reference`
+        :param options: a dict with options for transformations.
+        :type options: Optional[Dict[str]]
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/transformations/arraynotation2arrayrange_trans.py:151]: Added.

File: src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+
+def test_get_array_bound(fortran_reader):
+    ''' Test that the get_array_bound utility works as expected '''
+    # known bounds
+    psyir = fortran_reader.psyir_from_source(CODE)
+    node = psyir.walk(Reference)[0]
+    symbol = node.symbol
+    lower_bound, upper_bound, step = \
+        ArrayNotation2ArrayRangeTrans._get_array_bound(symbol, 0)
+    from psyclone.psyir.nodes import Literal, BinaryOperation
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py:89]: Oops. Moved to top.

File: src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author R. W. Ford
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py:34]: colonoscopied

File: src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+    assert step.value == "1"
+    # unknown bounds
+    psyir = fortran_reader.psyir_from_source(
+        CODE.replace("dimension(10)", "dimension(:)"))
+    node = psyir.walk(Reference)[0]
+    symbol = node.symbol
+    lower_bound, upper_bound, step = \
+        ArrayNotation2ArrayRangeTrans._get_array_bound(symbol, 0)
+    assert isinstance(lower_bound, BinaryOperation)
+    assert lower_bound.operator == BinaryOperation.Operator.LBOUND
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py:104]: I think I've done that.

File: src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+def test_validate():
+    ''' Test the validate method '''
+    trans = ArrayNotation2ArrayRangeTrans()
+    with pytest.raises(TransformationError) as info:
+        trans.validate(None)
+    assert ("The supplied node should be a Reference but found 'NoneType'."
+            in str(info.value))
+    with pytest.raises(TransformationError) as info:
+        trans.validate(Reference(DataSymbol("x", REAL_TYPE)))
+    assert ("The supplied node should be a Reference that references a "
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py:186]: Added.

File: src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py (hunk 2 above)
Comment (rupertford) [reply src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py:140]: Changed.

File: src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py (hunk 3 above)
Comment (rupertford) [reply src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py:108]: Added test for idx==1.

File: src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py (hunk 1 above)
Comment (rupertford) [reply src/psyclone/tests/psyir/transformations/arraynotation2arrayrange_trans_test.py:151]: Added new validate test based on this test. Kept this test as is. Let me know if you think it should be removed.

**Comment by rupertford** [comment]
I removed the #1576 todo in `fparser2.py` as it changes the code to use loop variables rather than ":" notation which is not relevant to what we are doing here.

**Comment by rupertford** [comment]
Ready for next review from @arporter

**Review by arporter — CHANGES_REQUESTED** [review]
All requested changes have been made.
UG has been extended and builds fine.
Ref. guide builds with no new warnings.
Tests and examples are all OK with compilation.

I would have merged but I've just read an email 
from Simon and that made me think about what happens when we have, e.g.:

    a%b = 0

where component `b` is an array. Does this transformation still work in that case?
If not, I think probably a TODO and an new Issue is in order, rather than prolong this one.

There's also a pylint error about an un-needed import in one of the test files.

**Comment by rupertford** [comment]
Added a test showing that an array within a structure causes the transformation to raise an exception. Also added a new issue #1858 about supporting arrays within structures and referenced this issue in the transformation.

**Comment by rupertford** [comment]
Removed unnecessary import from test file.

**Comment by rupertford** [comment]
Ready for next review from @arporter

File: src/psyclone/psyir/transformations/reference2arrayrange_trans.py [hunk 14]
```diff
@@ -133,6 +133,7 @@ def validate(self, node, options=None):
             symbol.
 
         '''
+        # TODO issue #1858. Add support for structures containing arrays.
```
Comment (arporter) [inline src/psyclone/psyir/transformations/reference2arrayrange_trans.py:136]: Sorry to be picky but I think we should mention this limitation in the docstring of the transformation (plus the associated issue number).

File: src/psyclone/tests/psyir/transformations/reference2arrayrange_trans_test.py [hunk 15]
```diff
@@ -218,6 +217,26 @@ def test_validate_range(fortran_reader):
            "'ArrayReference'." in str(info.value))
 
 
+def test_validate_structure(fortran_reader):
+    '''Test that a StructureReference raises an exception.'''
```
Comment (arporter) [inline src/psyclone/tests/psyir/transformations/reference2arrayrange_trans_test.py:221]: Please say that this is a limitation that will be tackled by #1858.

**Review by arporter — CHANGES_REQUESTED** [review]
Sorry to bounce it back yet again but I just want the current limitation to be better documented.
As a slight carrot, I've brought the branch to master and the tests and examples are still all fine.

File: src/psyclone/psyir/transformations/reference2arrayrange_trans.py (hunk 14 above)
Comment (rupertford) [reply src/psyclone/psyir/transformations/reference2arrayrange_trans.py:136]: Added to the docstring.

File: src/psyclone/tests/psyir/transformations/reference2arrayrange_trans_test.py (hunk 15 above)
Comment (rupertford) [reply src/psyclone/tests/psyir/transformations/reference2arrayrange_trans_test.py:221]: Added.

**Comment by rupertford** [comment]
I've hopefully made the required changes. Ready for another look from @arporter.
